### ISSUE TITLE
Cannot convert from 'System.Collections.Generic.IReadOnlyList<StrawberryShake.EntityIdOrData>' to ...

### ISSUE THREAD

**Issue by rafagp-dev** [ChilliCream/hotchocolate#4662]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

This is thrown when is trying to generate code:

> StrawberryShake.CodeGeneration.CSharp.Analyzers/StrawberryShake.CodeGeneration.CSharp.Analyzers.CSharpClientGenerator/GraphqlClient.StrawberryShake.cs(5463,115): error CS1503: Argument 1: cannot convert from 'System.Collections.Generic.IReadOnlyList<StrawberryShake.EntityIdOrData>' to 'System.Collections.Generic.IReadOnlyList<GraphQL.IClosePayment_CommitPayments_Results>

### Steps to reproduce

query:
```graphql
mutation ClosePayment($commitPaymentInput: [CommitPaymentInput!]!) {
  commitPayments(payments: $commitPaymentInput) {
    results {
      id
    }
  }
}
```

schema:
```graphql
...
union PaymentActionResult = ClaimPayment | PolicyPayment | PaymentCommitFailure | AccessDeniedError

type PaymentActionsResultList {
  results: [PaymentActionResult!]!
}
...
```

### Additional Context?

I believe the error is on the results list, but have no idea why that happens.

### Product

Strawberry Shake

### Version

12.5.0.0

**Comment by michaelstaib** [comment]
Can you post to me or pascal on slack (slack.chillicream.com) the full schema file?

**Comment by touriganational** [comment]
Hi, are you able to see the cause of this error?

**Comment by touriganational** [comment]
Hi, any update?

**Comment by michaelstaib** [comment]
The query is not valid GraphQL. 

```
results {
      id
}
```

Results is a union type and thus needs fragments to access any field.

**Comment by rafagp-dev** [comment]
> The query is not valid GraphQL.
> 
> ```
> results {
>       id
> }
> ```
> 
> Results is a union type and thus needs fragments to access any field.

I am sorry the query that I post was incorrect. This is the correct one:
```graphql
mutation ClosePayment($commitPaymentInput: [CommitPaymentInput!]!) {
  commitPayments(payments: $commitPaymentInput) {
    ...ClosePaymentResultList
  }
}

fragment ClosePaymentStaus on Payment {
  id
}

fragment ClosePaymentResultList on PaymentActionsResultList {
  results {
    ...ClosePaymentStaus
  }
}

```

**Comment by rafagp-dev** [comment]
> The query is not valid GraphQL.
> 
> ```
> results {
>       id
> }
> ```
> 
> Results is a union type and thus needs fragments to access any field.

I am sorry the query that I post was incorrect. This is the correct one:
```graphql
mutation ClosePayment($commitPaymentInput: [CommitPaymentInput!]!) {
  commitPayments(payments: $commitPaymentInput) {
    ...ClosePaymentResultList
  }
}

fragment ClosePaymentStaus on Payment {
  id
}

fragment ClosePaymentResultList on PaymentActionsResultList {
  results {
    ...ClosePaymentStaus
  }
}

```

**Comment by rhyshamilton** [comment]
@rafagp-dev or @touriganational, were either of you able to resolve this?

Setting "noStore": true` in the `.graphqlrc.json` seems to fix this for me, but doing does away with any caching benefits. Not sure if it's a long-term solution or if either of you had a more full-proof one?

**Comment by rafagp-dev** [comment]
Nope. We are using other field that dont use a list of unions for now, since that data is not too relevant for the moment in our project.

**Comment by gt4Rhys** [comment]
Just incase this helps anyone, I was able to solve this issue.

After a fair bit of debugging, I was able to figure out that I hadn't [explicitly mapped the IdType](https://chillicream.com/docs/hotchocolate/defining-a-schema/scalars.) for my id field on one of the objects in my union.

e.g - `descriptor.Field(x => x.Id) .Type<NonNullType<IdType>>();`
